### PR TITLE
feat: script distribute and run

### PR DESCRIPTION
**Pull request by nic-chen**
### What this PR does / why we need it:
support script distribute and run that can do a lot,  such as orchestrating plugins

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible?

### CODE REVIEW

File: apisix/init.lua [hunk 1]
```diff
@@ -451,19 +452,25 @@ function _M.http_access_phase()
...
-    api_ctx.plugins = plugin.filter(route, plugins)
-
-    run_plugin("rewrite", plugins, api_ctx)
-    if api_ctx.consumer then
-        local changed
-        route, changed = plugin.merge_consumer_route(route, api_ctx.consumer)
-        if changed then
-            core.table.clear(api_ctx.plugins)
-            api_ctx.plugins = plugin.filter(route, api_ctx.plugins)
+    script.load_script(route, api_ctx)
```
Comment (membphis) [inline apisix/init.lua:455]: name style: `script.load`, `script.run` is enough and clearer

File: apisix/init.lua [hunk 2]
```diff
@@ -451,19 +452,25 @@ function _M.http_access_phase()
...
+    if api_ctx.script_obj then
+        script.run_script("access", api_ctx)
```
Comment (membphis) [inline apisix/init.lua:457]: Is it possible move `script.load_script(route, api_ctx)` to here?

File: apisix/script.lua [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    if not obj then
+        return api_ctx
+    end
+
+    core.log.info("script_obj", core.json.delay_encode(obj, true))
+
+    local phase_fun = obj[phase]
+    if phase_fun then
+    	pcall(phase_fun, api_ctx)
```
Comment (membphis) [inline apisix/script.lua:33]: we should catch the return value, it may be failing. we need to record the error message.

File: apisix/script.lua [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+
+function _M.load_script(route, api_ctx)
+	local script = route.value.script
+	if script == nil or script == "" then
+		return nil
+	end
+
+	local loadfun = loadstring(script)
```
Comment (membphis) [inline apisix/script.lua:16]: we should catch the return value, it may be failing. we need to record the error message.

Comment (membphis) [reply apisix/script.lua:16]: the second value is useful, eg: `"=route#id"` or `"=service#id"`

**Comment by membphis** [comment]
@nic-chen I think you can write some doc now.

**Comment by membphis** [comment]
@nic-chen we need to add more test cases about Admin API. And we need to check if the input script is valid Lua source code.

**Comment by nic-chen** [comment]
> @nic-chen I think you can write some doc now.

done.

File: doc/architecture-design.md [hunk 5]
```diff
@@ -216,6 +217,25 @@ Not all plugins have specific configuration items. For example, there is no spec
...
+
+`Script` represents a script that will be executed during the `HTTP` request/response life cycle.
+
+The `Script` configuration can be directly bound to the `Route`.
+
+`Script` and `Plugin` are mutually exclusive, and `Script` is executed first. This means that after configuring `Script`, the `Plugin` configured on `Route` will not be executed.
+
+In theory, you can write arbitrary Lua code in `Script`, or you can directly call existing plugins to reuse existing code.
+
+`Script` also has the concept of execution phase, supporting `access`, `header_filer`, `body_filter` and `log` stages. The system will automatically execute the code of the corresponding stage in the `Script` script in the corresponding stage.
```
Comment (membphis) [inline doc/architecture-design.md:230]: stages -> phase

File: doc/architecture-design.md [hunk 6]
```diff
@@ -216,6 +217,25 @@ Not all plugins have specific configuration items. For example, there is no spec
...
+
+```json
+{
+    ...
+    "plugins": "local core = require(\"apisix.core\")\nlocal _M = {}\nfunction _M.access(api_ctx)\n core.log.warn(\"hit access phase\")\nend \nfunction _M.header_filter(ctx)\n core.log.warn(\"hit header_filter phase\")\nend\nfunction _M.body_filter(ctx)\n core.log.warn(\"hit body_filter phase\") \nend\nfunction _M.log(ctx)\n core.log.warn(\"hit log phase\")\nend\nreturn _M"
```
Comment (membphis) [inline doc/architecture-design.md:235]: we need a simpler example, this one is difficult to read.

Comment (membphis) [reply doc/architecture-design.md:235]: Should `plugins` be `script`?

Comment (membphis) [reply doc/architecture-design.md:235]: and please update the Chinese version too.

**Comment by nic-chen** [comment]
> I think we need to add a test case: enable "plugins" and "scripts" at the same time.
> 
> We should throw an error and not allow them to work together.

done.

File: apisix/init.lua [hunk 7]
```diff
@@ -451,19 +452,25 @@ function _M.http_access_phase()
...
+            route, changed = plugin.merge_consumer_route(route, api_ctx.consumer)
+            if changed then
+                core.table.clear(api_ctx.plugins)
+                api_ctx.plugins = plugin.filter(route, api_ctx.plugins)
+            end
         end
+        run_plugin("access", plugins, api_ctx)
     end
-    run_plugin("access", plugins, api_ctx)
+
```
Comment (membphis) [inline apisix/init.lua:473]: remove this blank line

File: apisix/schema_def.lua [hunk 8]
```diff
@@ -441,6 +443,13 @@ _M.route = {
         {required = {"upstream", "uris"}},
         {required = {"upstream_id", "uris"}},
         {required = {"service_id", "uris"}},
+        {required = {"script", "uri"}},
+        {required = {"script", "uris"}},
+    },
+    ["not"] = {
```
Comment (membphis) [inline apisix/schema_def.lua:449]: that is the first time I know this way, it is cool ^_^

File: apisix/script.lua [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+local loadstring = loadstring
+
+
+local _M = {}
+
+
+function _M.load(route, api_ctx)
+    local script = route.value.script
+    if script == nil or script == "" then
+        return nil
```
Comment (membphis) [inline apisix/script.lua:28]: I think we should throw an error if there is no `script`.

File: apisix/script.lua [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    end
+
+    local loadfun = loadstring(script, "route#" .. route.value.id)
+    api_ctx.script_obj = loadfun()
+end
+
+
+function _M.run(phase, api_ctx)
+    local obj = api_ctx and api_ctx.script_obj
+
```
Comment (membphis) [inline apisix/script.lua:38]: remove this line

File: apisix/schema_def.lua (hunk 8 above)
Comment (membphis) [reply apisix/schema_def.lua:449]: but I do not find any test case for this case

Comment (nic-chen) [reply apisix/schema_def.lua:449]: > but I do not find any test case for this case

`TEST 58` in t/admin/routes.t  is for this case  : )

File: apisix/script.lua [hunk 11]
```diff
@@ -25,18 +25,23 @@ local _M = {}
 function _M.load(route, api_ctx)
     local script = route.value.script
     if script == nil or script == "" then
+        core.log.error("fail to load empty script")
```
Comment (membphis) [inline apisix/script.lua:28]: we should use `error` to throw an error

File: apisix/script.lua [hunk 12]
```diff
@@ -25,18 +25,23 @@ local _M = {}
...
         return nil
     end
 
     local loadfun = loadstring(script, "route#" .. route.value.id)
+    if not loadfun then
+        core.log.error("fail to load script:", script)
```
Comment (membphis) [inline apisix/script.lua:34]: ditto

File: apisix/script.lua [hunk 13]
```diff
@@ -25,18 +25,23 @@ local _M = {}
...
+    end
     api_ctx.script_obj = loadfun()
 end
 
 
 function _M.run(phase, api_ctx)
     local obj = api_ctx and api_ctx.script_obj
-
     if not obj then
+        core.log.error("have no script object to run")
```
Comment (membphis) [inline apisix/script.lua:44]: ditto

File: apisix/script.lua [hunk 14]
```diff
@@ -25,18 +25,23 @@ local _M = {}
 function _M.load(route, api_ctx)
     local script = route.value.script
     if script == nil or script == "" then
+        core.log.error("fail to load empty script")
         return nil
     end
 
     local loadfun = loadstring(script, "route#" .. route.value.id)
```
Comment (membphis) [inline apisix/script.lua:32]: we need to use a `lru` way to cache the loaded object, avoid to call `loadstring` for every request.

Comment (membphis) [reply apisix/script.lua:32]: anyway, we can fix this in a new PR.

Comment (membphis) [reply apisix/script.lua]: I created a new issue, we can fix this later. https://github.com/apache/apisix/issues/1995

**Pull request merged**